### PR TITLE
Migrate data mapper compiler extension using new compiler API

### PR DESCRIPTION
**Pull request by isuruRuhu**
## Purpose
https://github.com/ballerina-platform/ballerina-lang/issues/30633

## Goals
Migrate data mapper compiler extension to new compiler API

## Approach
Break the data mapper compiler extension to two parts.

Validation of the connection sample data on the client side
Training data generator on the server side.

With this PR migrated the data mapper compiler extension using the latest compiler API to do the sample data validation.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java [hunk 1]
```diff
@@ -333,6 +121,11 @@ public void pluginExecutionCompleted() {
         for (String moduleName : listOfModuleNames) {
             processSampleDataFiles(moduleName);
         }
+        if (!this.dataMapperLog.getDataMapperPluginDiagnostic().isEmpty()) {
+            for (Diagnostic diagnostic : this.dataMapperLog.getDataMapperPluginDiagnostic()) {
+                compilationAnalysisContext.reportDiagnostic(diagnostic);
+            }
```
Comment (pubudu91) [inline src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java:127]: I'm assuming `getDataMapperPluginDiagnostic()` returns a list? If so, there's no need for the if check here.

**Review by pubudu91 — APPROVED** [review]
Approving since I don't see any major issue with the migration

**Review by azinneera — APPROVED** [review]
LGTM for the initial migration. We can improve based on the comments. @isuruRuhu please create an issue to track the imporvement points.

File: src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java [hunk 2]
```diff
@@ -71,260 +67,52 @@
 import java.util.stream.Stream;
 
 /**
- * Compiler extension to generate sample JSON files for data mapper.
+ * An {@code AnalysisTask} that is triggered for data mapper.
  */
-public class DataMapperPlugin extends AbstractCompilerPlugin {
-    public static final String URL_ENCODED_COLON = "%3A";
+public class SampleDataAnalysisTask implements AnalysisTask<CompilationAnalysisContext> {
```
Comment (azinneera) [inline src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java:72]: I think we can use SyntaxNodeAnalysisContext and remove the node visitor.

Comment (azinneera) [inline src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java:72]: I think we can use SyntaxNodeAnalysisContext

File: src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java (hunk 1 above)
Comment (isuruRuhu) [reply src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java]: Removed the if check

File: src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java (hunk 2 above)
Comment (isuruRuhu) [reply src/main/java/org/ballerinax/datamapper/SampleDataAnalysisTask.java:72]: Will address in a future PR

**Pull request merged**